pennylane/transforms/diagonalize_measurements.py

```diff
@@ -148,48 +149,51 @@ def circuit(x):
         [expval(X(0) + Z(1)), expval(X(0) + 0.2 * Z(1)), var(Y(2) + X(0))]
     """
 
-    bad_obs_input = [
-        o for o in supported_base_obs if o not in {qml.X, qml.Y, qml.Z, qml.Hadamard, qml.Identity}
-    ]
+    # Convert supported_base_obs to a frozenset once for faster membership checks
+    supported_base_obs = frozenset(supported_base_obs)
+    allowed_obs = frozenset({qml.X, qml.Y, qml.Z, qml.Hadamard, qml.Identity})
+
+    bad_obs_input = supported_base_obs - allowed_obs
 
     if bad_obs_input:
         raise ValueError(
             "Supported base observables must be a subset of [X, Y, Z, Hadamard, and Identity] "
             f"but received {list(bad_obs_input)}"
         )
 
-    diagonalize_all = set(supported_base_obs).issubset(set(_default_supported_obs))
+    diagonalize_all = supported_base_obs.issubset(_default_supported_obs)
 
     if to_eigvals and not diagonalize_all:
         raise ValueError(
             "Using to_eigvals=True requires diagonalizing all observables to the "
             "measurement basis. Observables "
-            f"{set(supported_base_obs)-set(_default_supported_obs)} can't "
+            f"{supported_base_obs - _default_supported_obs} can't "
             "be supported when using eigvals."
         )
 
-    if (
-        all(m.obs.pauli_rep is not None for m in tape.measurements if m.obs is not None)
-        and diagonalize_all
-    ):
+    # Precompute measurement observables that have a Pauli representation
+    pauli_measurements = [
+        m.obs for m in tape.measurements if m.obs is not None and m.obs.pauli_rep is not None
+    ]
+
+    if pauli_measurements and diagonalize_all:
```

Comment:
The code in `pauli_measurements` here doesn't do the same thing as the original code. The updated version determines whether there are _any_ observables with `pauli_rep`. This is not the correct condition for using `_diaglonaize_all_pauli_obs` - only circuits where _all_ observables have the `pauli_rep` attribute will be able to use the `_diagonalize_all_pauli_obs`. 

This change causes a new error to be raised, for example in this case:

```
dev = qml.device("default.qubit", shots=10)

@qml.transforms.diagonalize_measurements
@qml.qnode(dev)
def circuit():
    qml.RX(1.2, 0)
    qml.RY(2.34, 1)
    return qml.expval(qml.X(0)), qml.expval(qml.Hadamard(1))
```

This actually should have shown up in the tests, so thanks for catching that! I would suggest adding a test that fails with this current suggestion (ensuring that this branch is properly tested), and then updating the implementation so that the test passes. 

It should work with something along the lines of:

```suggestion
    # measurements with observable with no pauli_rep are not compatible with the default method
    incompatible_measurements = [
        m for m in tape.measurements if m.obs is not None and m.obs.pauli_rep is None
    ]

    if not incompatible_measurements and diagonalize_all:
```

I'm not sure how that fairs in terms of performance compared to what is currently on `master`.

Reply:
Added the test and corrected the code everything should be above board now :)